keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2861 @@
+  - 'name'
+  - 'namespace'
+  - 'operation'
+  - 'userInfo'
+  - 'dryRun'
+  - 'options'
+- 'config' - configuration data of the policy configuration being validated
+
+See below "Reporting violations to Clients" for more detail about how the
+`spec.validations` field works and how violations are reported.
+
+##### Policy Configurations
+
+Policy configurations are the interface between `ValidatingAdmissionPolicy`
+authors and the administrators that configure the policies for clusters. The
+configurations allows policy definition authors to define how exactly a
+`ValidatingAdmissionPolicy` may be configured using a OpenAPIv3 structural
+schema.
+
+How these configuration CRDs are defined has not yet been decided.
+Alternatives are summarized here and discussed in more detail below:
+
+<<[UNRESOLVED jpbetz, liggitt, ?? ]>>
+Reach consensus on this.
+<<[/UNRESOLVED]>>
+
+| Design Alternative                                | Summary                        |
+| ------------------------------------------------- | ------------------------------ |
+| Duck Typed CRDs                                   | Require CRDs be defined with match field of correct structure |
+| `PolicyConfiguration` kind with config reference  | Match criteria is defined in builtin type which references a config  | 
+| OpenAPIv3 `$ref` in CRDs                          | Match criteria is defined by k8s as a OpenAPIv3 type and CRDs schemas are modified to be allowed to reference it |
+| `/matchRules` subresource                         | Use a subresource to enforce CRDs provide match criteria      |
+| `PolicyConfiguration` kind with config embedded   | Match criteria is defined in a kind, it embeds a config CRD that is freeform |
+| Generate CRDs                                     | Policy defintions contain inline OpenAPIv3 schema for config, CRD is generated with both match and config sections |
+
+###### Alternative: Duck Typed CRDs
+
+This is the alternative shown in the initial examples of this KEP.
+
+Policy authors write a CRD to define how each policy is configured. E.g.:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: replicalimit.rules.example.com
+  annotations:
+    admission.kubernetes.io/is-policy-configuration-definition: "true"
+spec:
+  group: example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                config:
+                  type: object
+                    maxReplicas:
+                      type: int
+  scope: Cluster
+  names:
+    kind: ReplicaLimit
+```
+
+The `admission.kubernetes.io/is-policy-configuration-definition` annotation
+means "inject the correct .spec.match during admission and keep it up to date
+with a controller". (Suggested by deads2k). This minimizes version skew if new
```

Comment:
Is SSA configured to treat this as atomic? CRD schema has been used as an example where you wouldn't want multiple authors in the past, lol.

Reply:
You're totally right, I confirmed on a local cluster that the `spec.versions` is atomic. That's inconvenient. I was naively hoping to have the managed field owners nicely separated by using a controller. 

Since SSA is no help here, is it acceptable supplement the schema via a mutating admission controller? The client that writes the schema then stays as the managed field owner, they can always safely omit the `match` part of the schema and it will be added by the mutating controller.

Reply:
If you wanted to do this option, I'd make this not atomic if possible.

But I really think you should not do this option, upgrading this looks extremely hard.

Reply:
If you want to do this option I would make the user include this, and barf at type-check time if your clause isn't included / isn't right.

Reply:
I think we have consensus to not do this option.